pkg/operation/botanist/dnsresources_test.go

```diff
@@ -0,0 +1,301 @@
+					InternalDomain: &garden.Domain{
+						Provider: internalProvider,
+					},
+				},
+				Logger: logrus.NewEntry(logger.NewNopLogger()),
+			},
+		}
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+	})
+
+	Describe("#DeployInternalDNSResources", func() {
+		It("should delete the DNSOwner, DNSProvider, and DNSEntry resources, and then deploy the DNSRecord resource if the feature gate is enabled", func() {
+			defer test.WithFeatureGate(gardenletfeatures.FeatureGate, features.UseDNSRecords, true)()
+			gomock.InOrder(
+				internalDNSOwner.EXPECT().Destroy(ctx),
+				internalDNSProvider.EXPECT().Destroy(ctx),
+				internalDNSEntry.EXPECT().Destroy(ctx),
+				internalDNSRecord.EXPECT().Deploy(ctx),
+				internalDNSRecord.EXPECT().Wait(ctx),
+			)
+			Expect(b.DeployInternalDNSResources(ctx)).To(Succeed())
+		})
+
+		It("should migrate and delete the DNSRecord resource, and then deploy the DNSOwner, DNSProvider, and DNSEntry resources if the feature gate is disabled", func() {
+			defer test.WithFeatureGate(gardenletfeatures.FeatureGate, features.UseDNSRecords, false)()
+			gomock.InOrder(
+				internalDNSRecord.EXPECT().Migrate(ctx),
+				internalDNSRecord.EXPECT().WaitMigrate(ctx),
+				internalDNSRecord.EXPECT().Destroy(ctx),
+				internalDNSRecord.EXPECT().WaitCleanup(ctx),
+				internalDNSOwner.EXPECT().Deploy(ctx),
+				internalDNSOwner.EXPECT().Wait(ctx),
+				internalDNSProvider.EXPECT().Deploy(ctx),
+				internalDNSProvider.EXPECT().Wait(ctx),
+				internalDNSEntry.EXPECT().Deploy(ctx),
+				internalDNSEntry.EXPECT().Wait(ctx),
+			)
+			Expect(b.DeployInternalDNSResources(ctx)).To(Succeed())
+		})
+	})
+
+	Describe("#DeployExternalDNSResources", func() {
+		It("should delete the DNSOwner, DNSProvider, and DNSEntry resources, and then deploy the DNSRecord resource if the feature gate is enabled", func() {
+			defer test.WithFeatureGate(gardenletfeatures.FeatureGate, features.UseDNSRecords, true)()
+			gomock.InOrder(
+				externalDNSOwner.EXPECT().Destroy(ctx),
+				externalDNSProvider.EXPECT().Destroy(ctx),
+				externalDNSEntry.EXPECT().Destroy(ctx),
+				externalDNSRecord.EXPECT().Deploy(ctx),
+				externalDNSRecord.EXPECT().Wait(ctx),
+			)
+			Expect(b.DeployExternalDNSResources(ctx)).To(Succeed())
+		})
+
+		It("should migrate and delete the DNSRecord resource, and then deploy the DNSOwner, DNSProvider, and DNSEntry resources if the feature gate is disabled", func() {
+			defer test.WithFeatureGate(gardenletfeatures.FeatureGate, features.UseDNSRecords, false)()
+			gomock.InOrder(
+				externalDNSRecord.EXPECT().Migrate(ctx),
+				externalDNSRecord.EXPECT().WaitMigrate(ctx),
+				externalDNSRecord.EXPECT().Destroy(ctx),
+				externalDNSRecord.EXPECT().WaitCleanup(ctx),
+				externalDNSOwner.EXPECT().Deploy(ctx),
+				externalDNSOwner.EXPECT().Wait(ctx),
+				externalDNSProvider.EXPECT().Deploy(ctx),
+				externalDNSProvider.EXPECT().Wait(ctx),
+				externalDNSEntry.EXPECT().Deploy(ctx),
+				externalDNSEntry.EXPECT().Wait(ctx),
```

Comment:
The `DNSOwner` should be created after the `DNSEntry`. With the current order it can happen (and happened on my test env), that the DNS record on the DNS server is deleted by the `dns-external` extension first, before it is recreated as soon as the `DNSEntry` has be created. This is only relevant, if the owner TXT DNS record is still present when switching back from `UseDNSRecords=true` to `UseDNSRecords=false`.
As a consequence, the Kube-apiserver may be unavailable via both internal and external domain name for several minutes depending on the negative DNS caching on the client side.

Reply:
This order comes from the existing implementation in `DeployExternalDNS`. I would rather not do something special here, since what I want to do is "deploy the external DNS resources", so I can't understand how the existing implementation can be wrong if it worked so far? Or is it really only wrong because the the owner TXT DNS record is still present? In this case, if we take care to delete this record when switching to `UseDNSRecords=true`, so that it's not present in the case you mentioned, everything would be ok, correct?

These owner TXT records should be deleted anyway. I plan to add the needed fix to the AWS provider (and all other providers). I intentionally don't want to do something special for them in the Gardener code and leave that to the providers, the picture in Gardener is quite complicated and hard to understand already.

Reply:
Correct, if you take care to delete the owner TXT records when switching to `UseDNSRecords=true`, everything should be ok.

Reply:
OK, then we'll take care to delete them in the corresponding provider extensions.